website/src/app/blog/sans-io/readme.mdx

```diff
@@ -0,0 +1,654 @@
+  src="/images/blog/sans-io/stun-binding-state-machine2.svg"
+  alt="A UML state diagram for a STUN binding request that is being refreshed every 5s."
+/>
+
+The eventloop also changed slightly. Instead of exiting once we know our public
+IP, we'll now loop until the user quits the program:
+
+```rust
+    loop {
+        if let Some(transmit) = binding.poll_transmit() {
+            socket.send_to(&transmit.payload, transmit.dst).await?;
+            continue;
+        }
+
+        let mut buf = vec![0u8; 100];
+
+        tokio::select! {
+            Some(time) = &mut timer => {
+                binding.handle_timeout(time);
+            },
+            res = socket.recv(&mut buf) => {
+                let num_read = res?;
+                binding.handle_input(&buf[..num_read], Instant::now());
+
+            }
+        }
+
+        timer.reset_to(binding.poll_timeout());
+
+        if let Some(address) = binding.public_address() {
+            println!("Our public IP is: {address}");
+        }
+    }
+```
+
+## The premise of sans-IO
+
+So far, all of this seems like a very excessive overhead for sending a few UDP
+packets back and forth. Surely, the 10 line example introduced at the start is
+preferable over this state machine and the eventloop! The example might be, but
+recall the debate around function colouring. In a code snippet without
+dependencies like the above example, using `async` seems like a no brainer and
+really easy. The problem arises once you want to bring in dependencies.
+Composing your functionality (i.e. policy) on top of those dependencies imposes
+their decisions around async vs blocking IO on you. Libraries like `str0m` or
+`quinn-proto` which are written in the sans-IO way don't do that. Instead, they
+are pure state machines and thus the decision about async vs blocking IO or
+which async runtime to use is deferred to the application.
+
+Freedom to use either blocking or non-blocking IO isn't the only benefit to
+this. sans-IO design also compose very well, tend to have very flexible APIs,
+are easy to test and play well with Rust's features. Let's explore these
+additional benefits one by one.
+
+### Easy composition
+
+Take another look at the API of `StunBinding`. The main functions exposed to the
+eventloop are: `handle_timeout`, `handle_input`, `poll_transmit` &
+`poll_timeout`. None of these are specific to the domain of STUN! Most network
+protocols can be implemented with these or some variation of them. As a result,
+it is very easy to compose these state machines together: Want to query 5 STUN
+servers for your public IP? No problem. Just make 5 `StunBinding`s and call them
+in order[^3].
+
+In the case of firezone, you can see this in the example of
+[`snownet`](https://github.com/firezone/firezone/tree/main/rust/connlib/snownet),
+a library that combines ICE & wireguard and thereby exposes "magic" IP tunnels
+that work in any network setup to the rest of the application.
```

Comment:
I am not sure I full understand what you are trying to say with this change. It is kind of a given that a library is agnostic towards its dependents otherwise you would have circular dependencies?

Reply:
Oh my point was just that the sentence there didn't seem to make sense `will work ... to the rest of the application` maybe missing a word?

Reply:
"exposes IP tunnels (that work in any network setup) to the rest of the application"

Is what I wanted to say but perhaps that grammer is too German :)